content/blog/2021-07-06-Blog-Rohit-PropTest.md

```diff
@@ -0,0 +1,208 @@
+
+### Main motivation for applying
+
+My work involved writing "Property-based tests for tremor-script" and some of the reasons for applying are:
+
+- It involved a new type of testing I never heard of
+- Be part of a sandbox project where I can learn and grow with the community
+- The concept of learning Erlang + Rust was very interesting to me and frankly out of my comfort zone, as a person used to Python and web development in general.
+
+### New concepts I learned specific to my work
+
+- Erlang and Rust
+    - My work mostly revolved around Erlang and a little Rust and I was completely new to this ecosystem, it didn't help to not find much resources or actively accessible community for Erlang.
+    - I took this as a challenge and went through [various resources to learn Erlang](https://github.com/diru1100/learn_erlang), functional programming in general and I was able to see why this Language was involved to do the task at handi, [my mentor](https://twitter.com/heinz_gies) is very passionate about Erlang and shared his thought-process, experience which helped me broaden my knowledge and how to approach any concept while learning something completely new.
+- Tremor-script
+    - It is an interpreted expression-oriented language designed for the filtering, extraction, transformation and streaming of structured data in a stream or event-based processing system which is explicitly turing incomplete used to easily write programs specific to Tremor which is written.
+    - It is written using Rust but tested using Erlang
+- Property based testing
+    - We have seen various types of testing approaches like Unit testing, Integration testing, End2End testing etc.
+    - The purpose of tests is to check if our code is failing anywhere and test the same with various inputs.
+    - Fig 1 shows different types of tests to easily understand which features are specialise in what
+
+        ![](/img/blog/LFC-blog-diru/tests-comparison-graph.png)
+
+        Fig 1. Showing various tests wrt Feature compilance and Input scope covered [1]
+
+    - Property based testing takes a new appraoch which has the right balance of randomness and examples. They also have this nice feature of called shrinking which shows a simple version of sample input which is failing your tests. As in all PropTest is not an ideal solution to use everywhere but it fits our use-case here i.e testing features of a custom-language.
+    - Some of the resources I kept below may help understand PropTest better
+
+### Property Based testing in tremor-script
+
+- PropTest is implemented using [EQC](http://quviq.com/documentation/eqc/)
+- Quickcheck is the original library written for Haskell to do property based testing ( similar to xUnit for unit-tests) and EQC is the Erlang version of it
+- Erlang quickcheck or EQC is thte version used here
+    - Implemented by QuviQ
+    - Free version is quickcheck mini
+- Used to do PropTest
```

Comment:
I would ask to avoid 'PropTest' it read like a specific product and adds more jargon which makes the post harder to understand, can we stick with spelling out property based testing